### PR TITLE
Add FLAVA

### PR DESCRIPTION
**Pull request by gmittal**
```
python -m training.main \
    --save-frequency 1 \
    --zeroshot-frequency 1 \
    --report-to tensorboard \
    --hf-tokenizer-name gpt2 \
    --train-data=~/flava_cc/Train_GCC-training_output.csv \
    --val-data=~/flava_cc/Validation_GCC-1.1.0-Validation_output.csv \
    --csv-img-key filepath \
    --csv-caption-key title \
    --imagenet-val=~/imagenet_validation \
    --warmup 10000 \
    --batch-size=64 \
    --lr=1e-3 \
    --wd=0.1 \
    --epochs=30 \
    --workers=8 \
    --model flava-debug
```

### CODE REVIEW

File: src/training/params.py
Comment (rom1504) [inline src/training/params.py:71]: What kind of hf dataset do you use ?

File: src/training/train.py
Comment (rom1504) [inline src/training/train.py:56]: Does it make sense for this loss to be completely distinct? It seems flava is adding on top of clip rather than being distinct

File: src/training/params.py
Comment (rom1504) [inline src/training/params.py:327]: This should be configured at the model type level rather than being a param

File: src/open_clip/flava_model.py
Comment (rom1504) [inline src/open_clip/flava_model.py:14]: Is it possible to instead use any visual encoder supported by openclip?

**Comment by rom1504** [comment]
Left a few comments

Overall looks a bit big but interesting

Are you interested in running some experiments at scale ( for example 32 epochs of laion400m) to make sure it works ? If yes we could provide compute

**Comment by rom1504** [comment]
Could you please rebase on main ?

File: src/training/train.py
Comment (iejMac) [reply src/training/train.py:56]: yeah, maybe we add all of the losses separately: MIM, MLM, ITM, MMM, GC (would be our standard ClipLoss) and as a parameter you pass in which of those you want to use. Then you could also try things like CLIP + MIM etc. And FLAVA would basically be passing in all of the possible losses

File: src/open_clip/flava_model.py
Comment (iejMac) [inline src/open_clip/flava_model.py:72]: maybe move to transformer.py

**Comment by iejMac** [comment]
if we decide to go the 'merge FLAVA methods' route instead of the 'add FLAVA separately' route, we could do something where you pass parameters for what pretraining objectives you want to include and then add some ObjectiveSampler class that uses the loss functions according to some schedule. In FLAVA they use a round-robin schedule but I'd be interested to see how different ones effect training

File: src/training/data.py
Comment (usuyama) [inline src/training/data.py:617]: Need to pass the new collate_fn for webadaset as well? (around  `dataloader = wds.WebLoader(`)

File: src/training/params.py
Comment (usuyama) [reply src/training/params.py:327]: I agree it's better to consolidate it to model configs, but currently context_length is not reflected to the tokenizer.
Perhaps, it's a separate work item for corpora with longer text.

File: src/training/train.py
Comment (usuyama) [reply src/training/train.py:56]: I agree! It'd be awesome if we can make it modular and plug-in those losses separately.
Losses like local (token-aware) contrastive loss + MLM are important for biomedical domains too (https://arxiv.org/abs/2204.09817).

File: src/training/params.py
Comment (iejMac) [reply src/training/params.py:327]: @usuyama I made an issue for this (https://github.com/mlfoundations/open_clip/issues/222). Let's discuss how to best do it there

File: src/open_clip/flava_model.py
Comment (apsdehal) [inline src/open_clip/flava_model.py:123]: Same as above.

File: src/open_clip/flava_model.py
Comment (apsdehal) [inline src/open_clip/flava_model.py:243]: Looks like you have two projections now?

File: src/open_clip/loss.py
Comment (apsdehal) [inline src/open_clip/loss.py:217]: Make the weights configurable?

File: src/open_clip/tokenizer.py
Comment (apsdehal) [inline src/open_clip/tokenizer.py:201]: Original FLAVA doesn't use anything special just BERT tokenizer. Would it make sense to add original FLAVA version first and then move on to customizations?

Comment (apsdehal) [reply src/open_clip/tokenizer.py:201]: Instead of doing it this way, I would suggest using HuggingFace hub to save the tokenizer with custom tokens and then loading the custom tokenizer directly. Then this custom code logic doesn't need to be added here which is anyways very specifically to experiment we will run.

File: src/training/data.py
Comment (apsdehal) [inline src/training/data.py:43]: Hmm, the original FLAVA model uses normal MLM for the text but whole word mask MLM for the multimodal MLM. We should maybe merge that later for both to use normal MLM. (Maybe leave it as the original for now).

File: src/training/data.py
Comment (apsdehal) [inline src/training/data.py:36]: Very specific to FLAVA logic. Maybe the collate function should take a custom preprocessing function that it applies and returns the output. Then, in your training script you can specific the custom collate function you want.

File: src/training/data.py
Comment (apsdehal) [inline src/training/data.py:81]: Yeah, this definitely needs to be encapsulated. Otherwise, this will become a mess soon.

File: src/training/data.py
Comment (apsdehal) [inline src/training/data.py:96]: Maybe make it more generic by call it text key?

File: src/training/data.py
Comment (apsdehal) [inline src/training/data.py:98]: Is there a typo here with tokenizer_name? You are assigning tokenize below? Maybe you want tokenizer instead everywhere?

File: src/training/data.py
Comment (apsdehal) [inline src/training/data.py:531]: This is returning a dataloader.

File: src/training/data.py
Comment (apsdehal) [inline src/training/data.py:538]: Like mentioned above, this should be taken as preprocess function to be passed to data collator.

File: src/training/train.py
Comment (apsdehal) [reply src/training/train.py:56]: Agreed. This can be decoupled into multiple losses and then FLAVALoss using those losses as modular plugins. It would be nice to test ablations as well.

File: src/training/train.py
Comment (apsdehal) [inline src/training/train.py:86]: This probably can't land as it is. We need a batch level abstraction which moves all of the elements in the batch directly to GPU when `.to('cuda')` is done. Otherwise, these kind of changes will bloat the trainer.

File: src/training/train.py
Comment (apsdehal) [inline src/training/train.py:116]: Then the batch here goes directly to the model as it is. The loss is returned and then passed down the line. This will make sure the trainer is simplified.

**Review by apsdehal — CHANGES_REQUESTED** [review]
Thanks for this PR! the initial version is looking good. There are a lot of FLAVA specific changes which might bloat the repo if we land them as it is. 
There would be some changes required in OpenCLIP itself to make it work with a generic structure. From that point onwards, the repo will start moving towards a framework-style architecture which is also when some repos/libs/framework lose their focus and become bloated unless it is done properly. 
This will be a pivotal moment for this repo and I would though defer to the maintainers of OpenCLIP on the final decision on how they want to handle this. We can also keep this WIP as we continue developing on top.

**Comment by rwightman** [comment]
Things are going to start getting out of hand w/ nature of the current train loop and adding support for different losses, needing to handle different masks, etc... 

When I was experimenting with gradient caching I decided it was best to add another layer of abstraction to the train loop. I wanted to avoid going 'full lightning' on it, but made a minimal 'TrainJig' ... could also call it a 'TrainTask' to pull optimizer setup, loss handling, task specific step logic out of the main loop.

I stopped working on grad caching because the complexity was not proving worthwhile, but might be good to bring that abstraction in now https://github.com/rwightman/open_clip/blob/grad_cache/src/training/train_jig.py

We could have a ClipTrainJig/TrainTask, FlavaTrainTask, etc... to get rid of if is_flava, etc in main loop, and allow other multi-modal train tasks to be added without further mayhem...

File: src/training/params.py
Comment (rwightman) [reply src/training/params.py:71]: also curious what hf datasets are being used? I'm not aware of any train datasets that would be fitting the FLAVA @ scale objective? Also HF Datasets needs some work before it's going to scale well to large multi-modal datasets for big cluster runs off shared storage. All of our datasets on clusters being used for training right now are in webdataset tars...

Comment (usuyama) [reply src/training/params.py:71]: I'm curious about this too! I tried HF dataset once for diffusion fine-tuning ([huggingface/diffusers ref](https://github.com/huggingface/diffusers/pull/356)), but it didn't really scale well + it was a bit hard to debug with Arrow format.

File: src/open_clip/loss.py
Comment (apsdehal) [inline src/open_clip/loss.py:141]: Probably pass the ignore index here explicitly as well?

File: src/open_clip/model_configs/clip-flava-med.json
Comment (apsdehal) [inline src/open_clip/model_configs/clip-flava-med.json:12]: Standard is 77 I think. Are you counting CLS separately?

File: src/training/finetune.py
Comment (apsdehal) [inline src/training/finetune.py:157]: Leftover? Do you want to in general check for these and make sure you don't land rouge code here?

File: src/training/finetune.py
Comment (apsdehal) [inline src/training/finetune.py:1]: This is specific to FLAVA. We probably shouldn't land this file into main until it is more generic.

File: src/training/params.py
Comment (apsdehal) [reply src/training/params.py:71]: @gmittal Can you reply here?

File: src/training/train.py
Comment (apsdehal) [inline src/training/train.py:56]: You can probably recursively call it on dicts as well. I am not sure what level of compliance you would want to go towards.

File: src/training/train.py
Comment (apsdehal) [inline src/training/train.py:219]: Maybe I am missing something here. Where are you adding rest of the losses to the total loss that are returned from FLAVA?

Comment (apsdehal) [reply src/training/train.py:219]: or do you plan to make this particular part more generic by allowing loss registry? How are you pretraining FLAVA currently?

**Review by apsdehal — COMMENTED** [review]
Thanks @gmittal. We are almost there. Some more comments.

**Pull request closed without merging**
